src/libutils/oglapphelpers/metalapp.mm

```diff
@@ -0,0 +1,399 @@
+
+#include <GL/glew.h>
+#include <GL/glut.h>
+
+#else
+
+#include <GL/glew.h>
+#include <GL/gl.h>
+#include <GL/glut.h>
+
+#endif
+
+#include <OpenColorIO/OpenColorIO.h>
+
+#include "metalapp.h"
+#include "msl.h"
+#include "glsl.h"
+#include "mtltexture.h"
+
+namespace OCIO_NAMESPACE
+{
+
+struct GraphicsContext
+{
+    GraphicsContext() = delete;
+    GraphicsContext(NSOpenGLContext* glContext, id<MTLDevice> metalDevice)
+    {
+        this->glContext = glContext;
+        this->metalDevice = metalDevice;
+    }
+    ~GraphicsContext()
+    {
+        if(glContext)
+            [glContext release];
+        glContext = nil;
+    }
+    
+    NSOpenGLContext* glContext;
+    id<MTLDevice>    metalDevice;
+};
+
+MetalApp::MetalApp(const char * winTitle, int winWidth, int winHeight)
+    : ScreenApp(winTitle, winWidth, winHeight)
+{
+    initContext();
+}
+
+void MetalApp::initContext()
+{
+    NSOpenGLContext* currentContext = [NSOpenGLContext currentContext];
+    
+    m_context = std::unique_ptr<GraphicsContext>(new GraphicsContext(currentContext, MTLCreateSystemDefaultDevice()));
+}
+
+MetalApp::~MetalApp()
+{
+    m_metalBuilder.reset();
+    m_glStateBound = false;
+}
+
+void MetalApp::initImage(int imgWidth, int imgHeight, Components comp, const float * image)
+{
+    std::vector<float> imageData;
+    if(comp == Components::COMPONENTS_RGB)
+    {
+        RGB_to_RGBA(image, 3 * imgWidth * imgHeight, imageData);
+        image = imageData.data();
+    }
+    
+    setImageDimensions(imgWidth, imgHeight, comp);
+    m_image = std::make_shared<MtlTexture>(m_context->metalDevice, imgWidth, imgHeight, image);
+    m_outputImage = std::make_shared<MtlTexture>(m_context->metalDevice, m_context->glContext, imgWidth, imgHeight, image);
+    m_glStateBound = false;
+}
+
+void MetalApp::updateImage(const float *image)
+{
+    std::vector<float> imageData;
+    if(getImageComponents() == Components::COMPONENTS_RGB)
+    {
+        RGB_to_RGBA(image, 3 * m_image->getWidth() * m_image->getHeight(), imageData);
+        image = imageData.data();
+    }
+    m_image->update(image);
+}
+
+void MetalApp::prepareAndBindOpenGLState()
+{
+    // A dummyShaderDesc is enough.
+    // The builder will only be used to build GL program
+    GpuShaderDescRcPtr dummyShaderDesc = GpuShaderDesc::CreateShaderDesc();
+    m_oglBuilder = OpenGLBuilder::Create(dummyShaderDesc);
+    
+    std::ostringstream main;
+        
+    main <<    std::endl
+            << "uniform sampler2DRect img;" << std::endl
+            << std::endl
+            << "void main()" << std::endl
+            << "{" << std::endl
+            << "    gl_FragColor = texture2DRect(img, gl_TexCoord[0].st * "
+            << "vec2(" << m_outputImage->getWidth() << ", " << m_outputImage->getHeight() << "));"
+            << std::endl
+            << "}" << std::endl;
+    
+    
+    glActiveTexture(GL_TEXTURE0);
+    glBindTexture(GL_TEXTURE_RECTANGLE_EXT, m_outputImage->getGLHandle());
```

Comment:
This texture is shared between gl and metal. The way metal works, we create pixel buffer and then take an OpenGL handle, and a Metal texture handle from it. The way metal demands us to use this texture in OpenGL is to use its as a rectangle texture. It can't be used as a normal texture2d.